### PR TITLE
Assign specific jobs to dedicated workers

### PR DESCRIPTION
**Pull request by temaniarpit27**
https://github.com/0xPolygonZero/zk_evm/issues/507

### CODE REVIEW

File: paladin-core/Cargo.toml [hunk 1]
```diff
@@ -17,8 +17,9 @@ rustdoc-args = ["--cfg", "docsrs"]
 name = "paladin"
 
 [dependencies]
-anyhow = { version = "1.0.75", features = ["backtrace"] }
 dotenvy = "0.15.7"
+anyhow = { version = "1.0.75", features = ["backtrace"] }
```
Comment (temaniarpit27) [inline paladin-core/Cargo.toml:21]: fix this

File: paladin-core/src/channel/queue.rs [hunk 2]
```diff
@@ -158,42 +159,36 @@ impl<
...
-            )
+            .declare_queue(&self.identifier, self.channel_type.into())
             .await?;
 
         Ok(queue.publisher())
     }
 
     /// Get a receiver for the underlying queue.
     async fn receiver<'a, T: Serializable + 'a>(&self) -> Result<Self::Receiver<'a, T>> {
+        let identifier: String = rand::thread_rng()
```
Comment (temaniarpit27) [inline paladin-core/src/channel/queue.rs:170]: check the best way to do this

File: paladin-core/src/channel/queue.rs [hunk 3]
```diff
@@ -206,19 +201,24 @@ where
...
             channel_type,
         })
     }
 
-    /// Issue a new channel, generating a new UUID as the identifier.
-    async fn issue(&self, channel_type: ChannelType) -> Result<(Uuid, Self::Channel)> {
-        let identifier = Uuid::new_v4();
+    /// Issue a new channel, generating a unique identifier.
+    async fn issue(&self, channel_type: ChannelType) -> Result<(String, Self::Channel)> {
+        let identifier: String = rand::thread_rng()
```
Comment (temaniarpit27) [inline paladin-core/src/channel/queue.rs:214]: same here

File: paladin-core/src/runtime/mod.rs [hunk 4]
```diff
@@ -113,22 +113,27 @@ type CoordinatedTaskChannel<'a, Op, Metadata> = (
...
     task_channel: DynamicChannel,
     serializer: Serializer,
     worker_emulator: Option<Vec<JoinHandle<Result<()>>>>,
     _marker: Marker,
 }
-const TASK_BUS_ROUTING_KEY: Uuid = Uuid::nil();
-const IPC_ROUTING_KEY: Uuid = Uuid::max();
+
+// TODO - This is a hack for now. We need to figure out a better way to handle
+const IPC_ROUTING_KEY: &str = "ffffffff-ffff-ffff-ffff-ffffffffffff";
```
Comment (temaniarpit27) [inline paladin-core/src/runtime/mod.rs:126]: check the best way for this

File: paladin-core/src/runtime/mod.rs [hunk 5]
```diff
@@ -643,12 +653,17 @@ impl WorkerRuntime {
             }
         });
 
+        let identifier: String = rand::thread_rng()
```
Comment (temaniarpit27) [inline paladin-core/src/runtime/mod.rs:656]: check the best way for this

File: paladin-core/src/common/mod.rs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+use rand::{distributions::Alphanumeric, Rng};
+
+pub fn get_random_routing_key() -> String {
+    rand::thread_rng()
+        .sample_iter(&Alphanumeric)
+        .take(10)
```
Comment (muursh) [inline paladin-core/src/common/mod.rs:6]: Is there a reason we need this so big?

Comment (temaniarpit27) [reply paladin-core/src/common/mod.rs:6]: no 
we can reduce it to maybe 5 or something

File: paladin-core/src/runtime/mod.rs [hunk 7]
```diff
@@ -684,33 +704,34 @@ impl WorkerRuntime {
...
+            if terminated_jobs.contains_key(&payload.clone().routing_key) {
+                trace!(routing_key = %payload.clone().routing_key, "skipping terminated job");
                 acker.nack().await?;
 
                 continue;
             }
 
-            let routing_key = payload.routing_key;
+            let routing_key = payload.clone().routing_key;
+            let routing_key_clone = routing_key.clone(); // Clone here
```
Comment (muursh) [inline paladin-core/src/runtime/mod.rs:715]: What's this comment for?

File: paladin-core/src/runtime/mod.rs [hunk 8]
```diff
@@ -684,33 +704,34 @@ impl WorkerRuntime {
...
 
-            let span = debug_span!("remote_execute", routing_key = %routing_key.simple());
+            let span = debug_span!("remote_execute", routing_key = %routing_key_clone);
             let execution_task = payload.remote_execute().instrument(span);
 
             // Create a future that will wait for an IPC termination signal.
             let ipc_sig_term = {
                 let mut ipc_sig_term_rx = ipc_sig_term_rx.clone();
+                let routing_key_clone = routing_key_clone.clone(); // Clone here
```
Comment (muursh) [inline paladin-core/src/runtime/mod.rs:723]: Same.

File: paladin-core/src/runtime/mod.rs [hunk 9]
```diff
@@ -113,22 +113,30 @@ type CoordinatedTaskChannel<'a, Op, Metadata> = (
...
     channel_factory: DynamicChannelFactory,
     task_channel: DynamicChannel,
     serializer: Serializer,
     worker_emulator: Option<Vec<JoinHandle<Result<()>>>>,
     _marker: Marker,
 }
-const TASK_BUS_ROUTING_KEY: Uuid = Uuid::nil();
-const IPC_ROUTING_KEY: Uuid = Uuid::max();
+
+const IPC_ROUTING_KEY: &str = "ffffffff-ffff-ffff-ffff-ffffffffffff";
```
[the remark this answers is not in the document]
Comment (temaniarpit27) [reply paladin-core/src/runtime/mod.rs:125]: the earlier ipc routing key was uuid::max(), which is this string. i have kept is same for now but maybe we can do something more readable

File: paladin-core/src/runtime/mod.rs (hunk 7 above)
Comment (temaniarpit27) [reply paladin-core/src/runtime/mod.rs:715]: sorry, copilot issues

File: paladin-core/src/runtime/mod.rs (hunk 8 above)
Comment (temaniarpit27) [reply paladin-core/src/runtime/mod.rs:723]: will remove

File: paladin-core/src/config/mod.rs [hunk 10]
```diff
@@ -40,6 +40,10 @@ pub struct Config {
     /// Provides the URI for the AMQP broker, if the AMQP runtime is selected.
     #[arg(long, help_heading = HELP_HEADING, env = "AMQP_URI", required_if_eq("runtime", "amqp"))]
     pub amqp_uri: Option<String>,
+
+    /// Provides the routing key for workers to listen on, if the AMQP runtime
```
Comment (muursh) [inline paladin-core/src/config/mod.rs:44]: `if the runtime...`? It feels like something is missing

Comment (temaniarpit27) [reply paladin-core/src/config/mod.rs:44]: fixed

**Review by muursh — APPROVED** [review]

**Pull request merged**